modules/pam_motd/pam_motd.c

```diff
@@ -97,9 +97,192 @@ static void try_to_display_directory(pam_handle_t *pamh, const char *dirname)
+    cat_str = (char *)malloc(sizeof(char) * (strlen(a_str) + strlen(b_str) + strlen("/") + 1));
+    if (cat_str == NULL) {
+	return NULL;
+    }
+
+    strcpy(cat_str, a_str);
+    strcpy(cat_str + strlen(a_str), "/");
+    strcpy(cat_str + strlen(a_str) + strlen("/"), b_str);
+
+    return cat_str;
+}
+
+static int compare_strings(const void * a, const void * b) {
+    const char *a_str = *(char **)a;
+    const char *b_str = *(char **)b;
+
+    if (a_str == NULL && b_str == NULL) {
+        return 0;
+    }
+    else if (a_str == NULL) {
+	return -1;
+    }
+    else if (b_str == NULL) {
+	return 1;
+    }
+    else {
+	return strcmp(a_str, b_str);
+    }
+}
+
+static int filter_dirents(const struct dirent *d)
+{
+    return (d->d_type == DT_REG || d->d_type == DT_LNK);
+}
+
+static void try_to_display_directories_with_overrides(pam_handle_t *pamh, char **motd_dir_path_split, int num_motd_dirs)
+{
+    if (pamh == NULL || motd_dir_path_split == NULL) {
+	return;
+    }
+    if (num_motd_dirs < 1) {
+	return;
+    }
+
+    struct dirent ***namelist = NULL;
+    int *namelist_sizes = NULL;
+    char **namelist_final = NULL;
+
+    if ((namelist = (struct dirent ***)malloc(sizeof(struct dirent **) * num_motd_dirs)) == NULL) {
+	return;
+    }
+    if ((namelist_sizes = (int *)malloc(sizeof(int) * num_motd_dirs)) == NULL) {
+	return;
+    }
+
+    int total_size = 0;
+    for (int i = 0; i < num_motd_dirs; i++) {
+	namelist_sizes[i] = scandir(motd_dir_path_split[i], &(namelist[i]), filter_dirents, alphasort);
+	if (namelist_sizes[i] < 0) {
+	    total_size = 0;
+	    break;
+	}
+	total_size += namelist_sizes[i];
+    }
+
+    if (total_size == 0) {
+	return;
+    }
+
+    // allocates more than needed here, by allocatng space for duplicate names which are later skipped
+    if ((namelist_final = (char **)malloc(sizeof(char *) * total_size)) == NULL) {
+	return;
+    }
+
+    for (int i = 0; i < total_size; i++) {
+	namelist_final[i] = NULL;
+    }
+
+    int num_files = 0;
+    for (int i = 0; i < num_motd_dirs; i++) {
+	for (int j = 0; j < namelist_sizes[i]; j++) {
+	    namelist_final[num_files] = namelist[i][j]->d_name;
+	    num_files++;
+	}
+    }
+
+    qsort(namelist_final, total_size, sizeof(const char *), compare_strings);
+
+    for (int i = 0; i < total_size; i++) {
+	if (namelist_final[i] == NULL) {
+	    continue;
+	}
+	if (i > 0 && strcmp(namelist_final[i], namelist_final[i - 1]) == 0) {
+	    continue;
+	}
+
+	for (int j = 0; j < num_motd_dirs; j++) {
+	    char *abs_path = NULL;
+
+	    abs_path = join_dir_strings(motd_dir_path_split[j], namelist_final[i]);
+	    if (abs_path != NULL) {
+		int fd = open(abs_path, O_RDONLY, 0);
+		if (fd >= 0) {
+		    try_to_display_fd(pamh, fd);
+		    close(fd);
+
+		    // abs_path exists; do not continue looking in directories in motd_dir_path_split
+		    break;
+		}
+	    }
+	    _pam_drop(abs_path);
+	}
+    }
+
+    _pam_drop(namelist_final);
+    _pam_drop(namelist);
+}
```

Comment:
You're leaking here members of namelist array, the individual dirent entries, and also the namelist_sizes array.

Reply:
Sorry I missed that - addressed now.